File: problems/341_flatten_nested_list_iterator/solution.py
class NestedIterator:
    def __init__(self, nestedList: list):
        self.stack = []
        self._flatten(nestedList)
        self.idx = 0

    def _flatten(self, nested):
        for item in nested:
            if item.isInteger():
                self.stack.append(item.getInteger())
            else:
                self._flatten(item.getList())

    def next(self) -> int:
        val = self.stack[self.idx]
        self.idx += 1
        return val

    def hasNext(self) -> bool:
        return self.idx < len(self.stack)

File: problems/341_flatten_nested_list_iterator/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class NestedIterator {
    vector<int> list;
    int idx = 0;
    void flatten(vector<NestedInteger>& nestedList) {
        for (auto& ni : nestedList) {
            if (ni.isInteger()) list.push_back(ni.getInteger());
            else flatten(ni.getList());
        }
    }
public:
    NestedIterator(vector<NestedInteger>& nestedList) { flatten(nestedList); }
    int next() { return list[idx++]; }
    bool hasNext() { return idx < list.size(); }
};
